### PR TITLE
feat(go/adbc/driver/databricks) Schema Parser

### PR DESCRIPTION
**Pull request by jasonlin45**
# Summary
This change adds full support for parsing types from the raw type text when they come back from SQL warehouses in Databricks.

All of the data types as specified on this page should be supported:
https://docs.databricks.com/aws/en/sql/language-manual/sql-ref-datatypes

I've also adjusted the behavior to return a null column when we fail to parse so we can positionally propagate metadata to the IPC types.

# Demo

https://github.com/user-attachments/assets/af15516b-6b2b-413a-93ba-23a7893adccd

### CODE REVIEW

File: go/adbc/driver/databricks/schema.go [hunk 1]
```diff
@@ -52,6 +60,58 @@ var basicTypeToArrowTypeMap = map[sql.ColumnInfoTypeName]arrow.DataType{
...
+	// Simple types
+	"STRING":  arrow.BinaryTypes.String,      // character string values
+	"BINARY":  arrow.BinaryTypes.Binary,      // byte sequence values
+	"BOOLEAN": arrow.FixedWidthTypes.Boolean, // Boolean values
+	"VOID":    arrow.Null,                    // untyped NULL - not supported by Delta Lake
+	"NULL":    arrow.Null,                    // untyped NULL - not supported by Delta Lake
+}
+
+// tracks bracket and parenthesis depth for parsing nested types
+type depthTracker struct {
```
Comment (VersusFacit) [inline go/adbc/driver/databricks/schema.go:91]: Canonical solution is to use stacks for tracking as they'll also handle nesting. Less error prone. Easier to abstract over. There's even a LeetCode for it 😝

File: go/adbc/driver/databricks/schema.go [hunk 2]
```diff
@@ -128,182 +188,267 @@ func ResultSchemaToArrowSchema(dbxSchema *sql.ResultSchema) (*arrow.Schema, erro
 	return arrow.NewSchema(fields, &metadata), nil
 }
 
-func getArrowTypeFromColumnInfo(col sql.ColumnInfo) (arrow.DataType, error) {
-	if arrowType, ok := basicTypeToArrowTypeMap[col.TypeName]; ok {
+// Basic recursive descent parsing
+func getArrowTypeFromStringType(col string) (arrow.DataType, error) {
+	// handle simple, non-recursive types
```
[the remark this answers is not in the document]
Comment (VersusFacit) [reply go/adbc/driver/databricks/schema.go:193]: Also forgive me but this is hard to read / reason about as is -- encapsulating subcases into their own functions would be an immediate upgrade for me as a reviewer 🙇‍♀️

[the remark this answers is not in the document]
Comment (jasonlin45) [reply go/adbc/driver/databricks/schema.go]: I was taking a look through - and honestly each branch we check is at most 20 lines since we split out to helpers a lot already. I'm not sure if jumping around the code is easier to read. I think the git diff looks a bit ugly tbh

File: go/adbc/driver/databricks/schema.go (hunk 1 above)
Comment (jasonlin45) [reply go/adbc/driver/databricks/schema.go:88]: Resolved here: [53e83ac](https://github.com/dbt-labs/arrow-adbc/pull/27/commits/53e83ac1c0e379f62899f100bc0f2a4b4f629242)

It's checking correctness now so I got rid of the greedy regexes

File: go/adbc/driver/databricks/schema.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (VersusFacit) [reply go/adbc/driver/databricks/schema.go]: I can respect that 🖖 I think that it's very much a tradeoff and perhaps the git diff makes it especially bad where jumping around as you say is indeed no better

File: go/adbc/driver/databricks/schema.go [hunk 3]
```diff
@@ -29,6 +29,11 @@ import (
 
 const DbxSchemaTypeText = "type_text"
 
+const (
+	decimalTypeRegex  = `^(?:DECIMAL|DEC|NUMERIC)\((\d+)(?:,(\d+))?\)$`
+	intervalTypeRegex = `^INTERVAL\s+(YEAR(?:\s+TO\s+MONTH)?|MONTH|DAY(?:\s+TO\s+(HOUR|MINUTE|SECOND))?|HOUR(?:\s+TO\s+(MINUTE|SECOND))?|MINUTE(?:\s+TO\s+SECOND)?|SECOND)`
```
Comment (VersusFacit) [inline go/adbc/driver/databricks/schema.go:34]: This is specifically the SQL engine which does normalize upward, but one edge case here is to use case insensitive matching for protective coverage. May _not_ be needed though. It's probably worth just double checking if any code paths in this driver reach for a thrift env or similar, as those would require upcasing.

File: go/adbc/driver/databricks/schema.go [hunk 4]
```diff
@@ -128,182 +194,287 @@ func ResultSchemaToArrowSchema(dbxSchema *sql.ResultSchema) (*arrow.Schema, erro
...
-		// parse from "ARRAY<elementType>"
-		if matches := regexp.MustCompile(`ARRAY<(.+)>`).FindStringSubmatch(col.TypeText); matches != nil {
-			elementType := strings.TrimSpace(matches[1])
-			// Create a temporary ColumnInfo to recursively parse the element type
-			elementCol := sql.ColumnInfo{
-				TypeName: sql.ColumnInfoTypeName(elementType),
-			}
+	}
+	// { DECIMAL | DEC | NUMERIC } [ (  p [ , s ] ) ]
+	if matches := regexp.MustCompile(decimalTypeRegex).FindStringSubmatch(col); matches != nil {
```
Comment (VersusFacit) [inline go/adbc/driver/databricks/schema.go:231]: I don't know enough about go. Does this recompile them each time ? In this kind of subsystem compiling them once ahead of time _is_ worth it.

File: go/adbc/driver/databricks/schema.go [hunk 5]
```diff
@@ -128,182 +194,287 @@ func ResultSchemaToArrowSchema(dbxSchema *sql.ResultSchema) (*arrow.Schema, erro
...
+		var err error
+		var precision64 int64
+		precision64, err = strconv.ParseInt(matches[1], 10, 8)
```
Comment (VersusFacit) [inline go/adbc/driver/databricks/schema.go:234]: Magic numbers are annoying to read. make these into named consts please.

File: go/adbc/driver/databricks/schema.go [hunk 6]
```diff
@@ -128,182 +194,287 @@ func ResultSchemaToArrowSchema(dbxSchema *sql.ResultSchema) (*arrow.Schema, erro
...
+		if err != nil {
+			return nil, fmt.Errorf("invalid decimal precision: %v", err)
+		}
+		precision := int(precision64)
 
-			elementArrowType, err := getArrowTypeFromColumnInfo(elementCol)
+		var scale = 0
+		if len(matches) > 2 && matches[2] != "" {
```
Comment (VersusFacit) [inline go/adbc/driver/databricks/schema.go:241]: So this is saying it has a scale but you're passing the burden of this knowledge onto the reader which feels like code smell. Worst call add a comment this is explicitly for handling scale

File: go/adbc/driver/databricks/schema.go [hunk 7]
```diff
@@ -128,182 +194,287 @@ func ResultSchemaToArrowSchema(dbxSchema *sql.ResultSchema) (*arrow.Schema, erro
...
+	// If we get here, brackets were unbalanced
+	return "", false
+}
+
+func getArrowTypeFromStringType(col string) (arrow.DataType, error) {
+	// simple types - base case
+	if arrowType, ok := stringTypeToArrowTypeMap[col]; ok {
+		return arrowType, nil
+	}
+	if col == "TIMESTAMP" || col == "TIMESTAMP_NTZ" {
```
Comment (VersusFacit) [inline go/adbc/driver/databricks/schema.go:224]: why is timestamp logic here? I know it's not always worth jumping around but `if col == "TIMESTAMP" || col == "TIMESTAMP_NTZ" {` feels more general even if it's only ever used once. I don't see the logic in only having it here except to be DRY

File: go/adbc/driver/databricks/schema.go [hunk 8]
```diff
@@ -128,182 +188,267 @@ func ResultSchemaToArrowSchema(dbxSchema *sql.ResultSchema) (*arrow.Schema, erro
...
+		valueArrowType, err := getArrowTypeFromStringType(valueType)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse map value type: %w", err)
+		}
+		return arrow.MapOf(keyArrowType, valueArrowType), nil
+	}
+	// INTERVAL types - comprehensive pattern for all interval types
+	if matches := regexp.MustCompile(intervalTypeRegex).FindStringSubmatch(col); matches != nil {
+		// Year-month intervals
+		if strings.Contains(matches[1], "YEAR") || matches[1] == "MONTH" {
```
[the remark this answers is not in the document]
Comment (VersusFacit) [reply go/adbc/driver/databricks/schema.go:279]: Makes sense. It's hard to write good C-code (GO is just fake C) that is self documenting but I prompt it where possible. 🙏

File: go/adbc/driver/databricks/schema.go [hunk 9]
```diff
@@ -128,182 +194,287 @@ func ResultSchemaToArrowSchema(dbxSchema *sql.ResultSchema) (*arrow.Schema, erro
...
-
-			// fields are comma separated, nested types are in parentheses
-			fieldRegex := regexp.MustCompile(`([^,()]+(?:\([^()]*\)[^,()]*)*)`)
-			matches := fieldRegex.FindAllString(fieldsStr, -1)
-			for _, field := range matches {
-				fields = append(fields, strings.TrimSpace(field))
-			}
+		return arrow.StructOf(arrowFields...), nil
+	}
+	// default to an unrecognized type so we can still positionally propogate metadata
```
Comment (VersusFacit) [inline go/adbc/driver/databricks/schema.go:309]:
```suggestion
    // Use Null to occupy schema column position, so metadata remains correctly aligned even when the type is unrecognized.
```

maybe this? words har.

File: go/adbc/driver/databricks/schema.go [hunk 10]
```diff
@@ -128,182 +194,287 @@ func ResultSchemaToArrowSchema(dbxSchema *sql.ResultSchema) (*arrow.Schema, erro
...
+	// Split by whitespace to find modifiers
+	parts := strings.Fields(rest)
+	if len(parts) == 0 {
+		return "", true
+	}
+
+	// The first part is the base type, but it might contain nested parentheses
+	// We need to find where the type ends and modifiers begin
+	typeEnd := findTypeEnd(rest)
+	fieldType = strings.TrimSpace(rest[:typeEnd])
```
Comment (VersusFacit) [inline go/adbc/driver/databricks/schema.go:387]: eh, nit:

```
if fieldName == "" || rest == "" {
    return "", "", true, fmt.Errorf("invalid struct field: missing name or type")
}
```

Could check for correctness

File: go/adbc/driver/databricks/schema_test.go
Comment (VersusFacit) [inline go/adbc/driver/databricks/schema_test.go:1]: Just because I know this from my own test writing: Did you run these locally to see them pass?

**Review by VersusFacit — COMMENTED** [review]
Code quality and edge case minded comments. I left a LOT here but they're not really logic complaints, more defensive coding prompts. You've done a ton of the work already to chase out edge cases + I know there's an expectation that dbx SHOULD be normalized and correct.

Still a bit of defensive driving right now feels worth it since we have tools to speed up integration of that checking

File: go/adbc/driver/databricks/schema.go [hunk 11]
```diff
@@ -128,182 +194,287 @@ func ResultSchemaToArrowSchema(dbxSchema *sql.ResultSchema) (*arrow.Schema, erro
...
+}
+
+func getArrowTypeFromStringType(col string) (arrow.DataType, error) {
+	// simple types - base case
+	if arrowType, ok := stringTypeToArrowTypeMap[col]; ok {
+		return arrowType, nil
+	}
+	if col == "TIMESTAMP" || col == "TIMESTAMP_NTZ" {
 		return &arrow.TimestampType{
 			Unit:     arrow.Microsecond,
```
[the remark this answers is not in the document]
Comment (jasonlin45) [reply go/adbc/driver/databricks/schema.go:226]: this matches what they return via IPC

File: go/adbc/driver/databricks/schema.go (hunk 5 above)
Comment (jasonlin45) [reply go/adbc/driver/databricks/schema.go:234]: as in `BASE_10` and `BITS_8`?

not sure if that's particularly useful - maybe I can make this a commented helper function instead since we do this a lot

File: go/adbc/driver/databricks/schema.go (hunk 6 above)
Comment (jasonlin45) [reply go/adbc/driver/databricks/schema.go:241]: how would you do it then?

File: go/adbc/driver/databricks/schema.go [hunk 12]
```diff
@@ -128,182 +194,287 @@ func ResultSchemaToArrowSchema(dbxSchema *sql.ResultSchema) (*arrow.Schema, erro
...
-				TypeName: sql.ColumnInfoTypeName(elementType),
-			}
+	}
+	// { DECIMAL | DEC | NUMERIC } [ (  p [ , s ] ) ]
+	if matches := regexp.MustCompile(decimalTypeRegex).FindStringSubmatch(col); matches != nil {
+		var err error
+		var precision64 int64
+		precision64, err = strconv.ParseInt(matches[1], 10, 8)
+		if err != nil {
+			return nil, fmt.Errorf("invalid decimal precision: %v", err)
```
[the remark this answers is not in the document]
Comment (jasonlin45) [reply go/adbc/driver/databricks/schema.go:236]: great point - but I don't think we'd ever see a non base 10 return. If we do error for that reason, `err` will print it out, just have the wrong message.

File: go/adbc/driver/databricks/schema.go [hunk 13]
```diff
@@ -52,6 +57,67 @@ var basicTypeToArrowTypeMap = map[sql.ColumnInfoTypeName]arrow.DataType{
...
+	"LONG":     arrow.PrimitiveTypes.Int64, // 8-byte signed integer
+	"BIGINT":   arrow.PrimitiveTypes.Int64, // 8-byte signed integer
+
+	// Binary floating point types
+	"FLOAT":  arrow.PrimitiveTypes.Float32, // 4-byte single-precision
+	"REAL":   arrow.PrimitiveTypes.Float32, // 4-byte single-precision
+	"DOUBLE": arrow.PrimitiveTypes.Float64, // 8-byte double-precision
+
+	// Date-time types
+	"DATE": arrow.FixedWidthTypes.Date32, // year, month, day without timezone
```
[the remark this answers is not in the document]
Comment (jasonlin45) [reply go/adbc/driver/databricks/schema.go:77]: we are, just below because those aren't constants

[the remark this answers is not in the document]
Comment (VersusFacit) [reply go/adbc/driver/databricks/schema.go:82]: Yeah I was confused on who owns those types in this module. Makes sense enough though!

File: go/adbc/driver/databricks/schema.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (VersusFacit) [reply go/adbc/driver/databricks/schema.go:240]: nit: I'd like a comment on that so we don't have to rely on memory :D

File: go/adbc/driver/databricks/schema.go (hunk 6 above)
Comment (VersusFacit) [reply go/adbc/driver/databricks/schema.go:253]:
```suggestion
                // Match scale if present
                if len(matches) > 2 && matches[2] != "" {
```

maybe this? Helps identify what these ops are doing
